### DISCUSSION TITLE
Multi-Version Compatibility

### DISCUSSION THREAD

**Discussion by exectails** [NoCode-NoLife/melia#179] [Discussions]
Supporting multiple versions of a game's client is difficult. It's a lot of work and it can be complicated. However, especially in a game that has changed as much as ToS has over the years, such functionality might be highly desired by users. Maybe you want a feature back that was removed, maybe you want to use a client from a region that isn't supported anymore, or maybe you just want to experience the game as it was during the beta. There are situations where modifying the client a little bit doesn't cut it, or where readding features would take a considerable amount of work. Not to mention that not every user has the technical knowhow to do these changes.

Originally, we just started emulating iToS because that would become our hometurf, and we were too busy getting the basics to work to worry about keeping multiple clients compatible. That being said, if we do want multi-version compatibility... it should be kept in mind rather sooner than later, because that decision will affect the design of several systems, and adding support for it later on might get messy.

**Advantages**

- Ability to (potentially) use any client you want, particularly one that might be closest to your ideal version of the game
- Ability to keep using one client version even if the server is kept getting updated

**Disadvantages**

- Work. A lot of work. And headaches. Ulcers... the usual.

This was already briefly discussed once in #47, but ultimately I decided against it at the time. That was pre-Rebuild though, and I'd like to point your attention at the first comment on that issue.

>I don't think this game will last longer than 4-5yrs. Not sure there will be some milestones like RO ones:
classic -> pre-re (12-13episode) -> renewal
>
>Even if in future there will be some gamebreaking feature, some limitations or stupid exploitable mechanic it's should be in emu configs (like limited trade: 0 or 1) etc.
>
>Point is - if this feature (multi-ver) will be ever used by some player base? probably not

The game did last longer than 5 years, it did already have one major update that is on the level of Renewal, and I would wager that quite a few users might be interested in going back to pre-Rebuild. Not to mention all the other smaller changes, and who knows what will come in the future.

**Comment by exectails** [comment]
If we did want this feature, I believe there are two fundamental ways to implement it in a viable manner with C# for this game.

**1) Ifs and lookup tables**

We do runtime checks for packet changes and dynamic lookups for enum values. Over the years I've come up with a way to handle enums that doesn't suck completely, which I'm using in other projects, and it works pretty well in my opinion. For example, here's what a packet sender might look like.

```cs
public static void ShowEffect(Character character, EffectId effectId, EffectParameters parameters)
{
	if (Game.Version < 20050311)
		throw new GameVersionException("Not supported by this version.");

	var networkEffectId = MEnum<EffectId>.GetValue(effectId);

	var packet = new Packet(Op.NET_SHOW_EFFECT, character.ObjectId);
	packet.PutInt(networkEffectId);

	if (Game.Version >= 20060802)
		parameters?.AddTo(packet);

	character.Region.Broadcast(packet, character);
}
```

The `EffectId` enum is just a simple enum without explicit enumeration, which works independently from the values expected by the client. Inside the sender, hidden from the coder, this value is then converted to the value that the client expects, similar to how `HostToNetworkOrder` might affect the data sent over the network. `MEnum` is initialized from a file for every enum it needs to support and then just converts the values back and forth. The files it loads the values from use a C-like, macro-esque, enum-focused layout.

```c
StopMusic
PetGaze = 1000
PetBuff
#if VERSION >= 20060801
	MagicalMusic
	PetPvp
#endif
Curtain
```

This is one way we could handle enums. In the context of ToS, we could look at the property ids for an example of how this might be used. In the normal code, a normal enum would be used, but when the properties are supposed to get sent to the client, we look up the value the property id _should_ have for the currently selected version and send that instead.

Maintaining the enums would require some effort, but I believe it's manageable, and throwing in a few simple ifs when adding new packet fields isn't that much work.

**2) Branches**

This option was suggested by Tachiorz back when we first discussed this. Basically, we would have compatibility branches for certain versions, so if you wanted kToS support for example, you might use the branch `compatible-kr-latest`, or something along those lines, and that branch would only be compatible to the latest version of kToS.

I imagine that with this approach, you might keep one branch that is relatively free from version dependent enums, and then create branches off of it, that add the version dependent files, like ops, properties, etc. Finally, you might make master the default compatibility branch for the latest iToS, since it will be the most-used version for us.

This approach would presumably be pretty clean, but it would require a lot of branch hopping and potentially confuse users and contributors. It would also let us only support so many versions, because we can't viably create dozens or hundreds of compatbility branches.

**Non-Viable**

Two options that aren't viable in my opinion are `#define` and OOP. C#'s defines aren't on the level of C's. We'd have to create defines like `VERSION_20210831` for every single version, and filtering which part of the code we actually want to be active would be a mess. Similarly, I believe any kind of OOP solution is out, because the game's packets and ops just change _way_ too much. We would end up with thousands of variations of packet handlers and writers, to a completely unmaintainable degree.

**Reply by ovflowd** [reply]
I'm thinking of a third viable option. What about a Plugin System? Imagine that Melia's Core is all the basics (Quest System, Battle System, Managers, Connection Handler, Game Map, Pathfinding, and all the core stuff), and then you actually have the Plugin System which actually loads "Features" of the Game.

For example, "TP Trade Shop" is an in-game feature from the Plugin System. Each Plugin is a DLL when built, placed inside the "Plugins" folder.

Plugins have a contract and can extend multiple features of the Emulator. The Emulator during bootstrap loads numerous kinds of plugins.

Plugins can define their own MessageHandlers, Features and even Extended system. So for example, one could make A Chess NPC Plugin. Or an altered version of the Market which extends the Default Market Plugin. (Basically, an Extended Market Plugin extends the Default Market Plugin source).

So all game features are independent of the main source code, but still the source fo the "main" plugins are inside the `melia/src/plugins` folder for example. But they are built independently from the main Melia Project (prj) but altogether in the (Sln) (Solution).

**Reply by Adephx** [reply]
Do you foresee the Melia project picking up enough traction to justify keeping up with multiple revisions though? It seems interesting, but I would suggest to keep patching up the emulator and keep up with the current official release, then if at some point a major update occurs, we can decide to fork and keep maintaining both versions. You said it yourself, that it is almost impossible to emulate features without the access to the live server response, wouldn't that make all of the revisions incomplete, unless we somehow manage to finish this one within the current update cycle?

**Reply by ovflowd** [reply]
>  It seems interesting, but I would suggest that keep patching up the emulator to keep up with the current official release

I believe that IMHO, keeping the emulator itself (the core) agnostic of client and in-game based features is fundamental. Regardless of their versions.

That's why every in-game feature should be a Plugin, and the Plugins have their own versions. Versioning the Plugins could work in a matter where major updates of the Plugin are related to the functionality of the Plugin, Minor versions are related to fixing bugs and patch versions to update the Plugin to a new release of the Client.

**So basically:**

```
MarketPlugin v1.0.RELEASE-i23093 => Version 1.0 of the plugin with patch version for the i23093 release
MarketPlugin v1.1.RELEASE-i123903 => Version 1.1 of the plugin with patch version for the i23093 release
MarketPlugin v2.0.RELEASE-i322222 => Version 2.0 of the plugin with patch version for the i32222 release (A new version)
MarketPlugin v2.0.RELEASE-i123903 => Version 2.0 of the plugin with retro-compatibility to the release i23093
```

**Reply by exectails** [reply]
A plugin system could be used for version dependent features and systems, extending classes, adding new packet handlers, etc. That's true. And it could keep the core a little cleaner, since we wouldn't include features that 9 out of 10 versions aren't even using. But I believe it doesn't solve the underlying issue of how to handle packet and enum changes, does it?

For example, the Market system didn't exist originally. It was added later, so it would be a good example of something that could be added via plugin on demand, based on the version. Inside that Market plugin will be packet handlers and senders however. Some enums might've changed over time as well. And the question is how to handle these changes, that occur all over the core _and_ potential plugins, depending on the version the user chooses.

**Reply by exectails** [reply]
>it is almost impossible to emulate features without the access to the live server response

Actually, it's by no means impossible, it just requires more effort, because you have to reverse engineer the packet structures. We also know that there is a leak out there, a pre-Rebuild one, which might offer us the oppurtunity to log older packets at some point.

**Reply by ovflowd** [reply]
> But I believe it doesn't solve the underlying issue of how to handle packet and enum changes, does it?

As I was saying, all the game features (everything that actually communicates through a packet) would be a plugin. The Plugins actually have the feature logic for a specific feature. This means Melia's core would only provide everything that you need to make a feature work.

Example: UserManager (To add an effect, change their names, etc), MarketPlugin (for making the market), and et cetera. Think that all the basic features of the server itself (Quests, NPCs, Pathfinding, Network, Database, Communication, Broadcast, User Management, GameMaps, Achievements and et cetera are all part of the core).

But the implementation (usage) of those Classes and Services are done by the Plugins.

So for example the Contract between the Emulator and Plugins works as that when the Emulator loads it will iterate through the Plugins, and the Plugins will basically say. "Hey Melia, please map this Packet to this method that I provide".

**Reply by ovflowd** [reply]
Example implementation in Java for a Habbo Emulator: https://git.krews.org/morningstar/Arcturus-Community/-/tree/master/src/main/java/com/eu/habbo/plugin

**Reply by ovflowd** [reply]
> For example, the Market system didn't exist originally. It was added later, so it would be a good example of something that could be added via a plugin on-demand, based on the version. Inside that Market, plugins will be packet handlers and senders, however.

I covered that exactly in my last comment. Even the Login of the User itself (Client-side to Server-side) would be done via a Plugin. In the sense, of course, Melia is the one that communicates and handles the Sockets, but the Plugins would actually do the "work".

So basically a "PlayerLoginPlugin.dll" would manage the Login of the Player. Yes, we would need to ensure that each Plugin is independent and also that the shared functionalities (eg.: AddEffect, UpdateTeamName, RemoveEffect, Shout, Whisper, etc) are actually part of Melia, but the Plugins are the ones that invoke those methods.

**Reply by Adephx** [reply]
> > it is almost impossible to emulate features without the access to the live server response
> 
> Actually, it's by no means impossible, it just requires more effort, because you have to reverse engineer the packet structures. We also know that there is a leak out there, a pre-Rebuild one, which might offer us the opportunity to log older packets at some point.

Was the leak source or compiled? I said almost impossible, I should have said a nightmare. In the grand scheme of things almost everything is possible.

**Reply by exectails** [reply]
But what about the actual version changes? Let's take the Market plugin and postulate, just for an extreme example, that the packet that is sent to the client for displaying the available items changes once a month, because IMC added some new feature.

```cs
packet.PutInt(itemId);
packet.PutInt(newField1); // added two months ago
packet.PutInt(newField2); // added last month
packet.PutInt(newField3); // added today
```

How are you going to add these new fields? Would you create a new version of the entire plugin every month? And how would you manage these dozens and hundreds of plugin versions? Would there just be a repo to choose a version from? What if you need to fix a bug in the system, and that needs to be reflected in all of those versions of the plugin? That's basically the OOP approach, just in separate DLLs. And the core, which of course isn't a plugin, will have the same problem, but it will change even more often. Like seriously, at times there will be packet and enum changes every week.

Or am I misunderstanding something and you're suggesting that _all_ the packet reading and writing happens within the core? But then the plugin system would seem unrelated to the topic at hand 🤔

**Reply by exectails** [reply]
> Was the leak source or compiled? I said almost impossible, I should have said a nightmare. In the grand scheme of things almost everything is possible.
As you have pointed out, I like a challenge, and I would have no qualms about digging into an older client to get the packets^^ Especially since the current version of the client is far away from my ideal for the game. But if we don't have any kind of multi-version support, that's not even an option.

The leak was supposedly both actually. I saw the binaries, but the seller wasn't willing to pick apart the source code files, to show me some, to confirm to me that the source was legit and of the version that he stated.

**Reply by ovflowd** [reply]
I suppose we should not try to support every freaking release but keep with major ones. Even because we just can't keep updating the client non-stop. We should only update our server based on major releases.

As I was saying the versioning schema would do that for you. In a certain sense, when we actually update the Plugin it is to introduce a new feature or a bug fix. But for client packet changes we should just update the packets. This would also imply that we would have Git tags for plugin versions. That's where rebasing comes into.

When I'm updating a new feature into the plugin and need to update all the "client" (patch) versions, you just rebase the new update unto the patch versions. It's of course more easily said... I believe we should support only a very handful amount of versions of the client.

Saying that we should adopt a policy for which client versions we update, think about Linux versioning for example. We could always support 3 versions. (RTS, LTS, LTT) for example. (The one that is the newest, the one that is almost-newest but more stable, and the LTS which is a long term one for many releases).

Just an idea.

**Reply by Adephx** [reply]
I think I speak for the most people here, when I say that everyone enjoys the old version of the game a lot more. Seeing that you are the main driving force behind this project (so far anyway) are you sure that we shouldn't skip the latest revisions and just focus on that instead? We can do a poll about who actually wants the new features.

If the leak really contains the source files, we wouldn't even need an emulator, as we could just alter the code, even though that would put us in a tough spot with legality of the project.

**Reply by ovflowd** [reply]
I completely understand that it seems complicated. And yes, each Plugin would have its MessageHandlers (Incoming Messages from Client) and then the MessageSenders (Server sending to Client), basically those parts of the Plugins would change.

Basically, the Services and Classes provided by the plugins (The actual application logic that handles the feature) would never change from version to version, only when it is really needed.

**Reply by exectails** [reply]
Okay, in that case it sounds like you're proposing the branching model I described, just with plugins added on top^^ Focus on major versions and use branches to keep logic and version dependent code organized. The advantages and disadvantages outlined above should apply then.

>you are the main driving force behind this project (so far anyway)

Well, as the original developer, I kind of get roped into working on Melia every time someone shows some interest in the project again, and I'd be happy to see it actually take off after all those years, but honestly, in its current form, Melia isn't high on my list of interesting projects anymore, because I have absolutely no interest in the current version of the game.

>are you sure that we shouldn't skip the latest revisions and just focus on that instead

I'm not sure about sticking to the latest version at all, and it's not what I would want personally, but frankly, working without packet logs is an ordeal. Not difficult, or almost impossible, it's just tedious, and it takes time. Ideally you also want to have a little bit of reverse engineering experience, to understand the client code. That's not exactly well suited for a community project, because it would hinge on someone digging up the packet structures that weren't logged in the past, wheras, with the latest version, someone can just log into iToS and get whatever they want.

This is in part why I opened this discussion, because with multi-version support and a solid foundation you can have both.

>If the leak really contains the source files, we wouldn't even need an emulator, as we could just alter the code, even though that would put us in a tough spot with legality of the project.

You said it, it would be illegal. Sharing the source code _or_ the binaries publically in any form would be kind of crazy. And I certainly wouldn't recommend setting up public servers with those files. They _could_ be useful for research however. But even if we assume that it's just a matter of time until they become available, we don't know when that might be yet.

**Reply by ovflowd** [reply]
> e. That's not exactly well suited for a community project, because it would hinge on someone digging up the packet structures that weren't logged in the past, wheras, with the latest version, someone can just log into iToS and get whatever they want.

Reverse Engineering can be way harder and would require way more commitment with the contributors. I love reverse engineering. But we would need a specialist that understands that. I have friends that might be willing to volunteer for that. Yet, I definitely agree just using a Packet Logger is way easier because you get what you want.

**Reply by Adephx** [reply]
Just a reminder that ToS doesn't really recieve updates in a traditional sense. It should be relatively easy to convert the current state of the game back to what it was, even with the current client. Drops are handled on the server side, so are stats and access to skills.

**Reply by exectails** [reply]
>It should be relatively easy to convert the current state of the game back to what it was

Drops, stats, and skills surely are part of it, but there are always things you can't just easily revert back. But if I understand you correctly, you don't mind much either way, new version, old version, as long as the base systems are working in the end, right? Though you seem to be against multi version support, because it adds a lot more work that might never be put to good use.

**Reply by Adephx** [reply]
> > It should be relatively easy to convert the current state of the game back to what it was
> 
> Drops, stats, and skills surely are part of it, but there are always things you can't just easily revert back. But if I understand you correctly, you don't mind much either way, new version, old version, as long as the base systems are working in the end, right? Though you seem to be against multi version support, because it adds a lot more work that might never be put to good use.

That would be the TLDR of it yeah.
